crates/extensions/c8y_mapper_ext/src/entity_cache.rs

```diff
@@ -0,0 +1,229 @@
+use std::collections::HashMap;
+use tedge_api::entity::EntityExternalId;
+use tedge_api::entity::EntityMetadata;
+use tedge_api::entity_store::EntityRegistrationMessage;
+use tedge_api::entity_store::EntityType;
+use tedge_api::mqtt_topics::default_topic_schema;
+use tedge_api::mqtt_topics::EntityTopicId;
+use tedge_api::mqtt_topics::MqttSchema;
+use tedge_api::pending_entity_store::PendingEntityData;
+use tedge_api::pending_entity_store::PendingEntityStore;
+use tedge_mqtt_ext::MqttMessage;
+use thiserror::Error;
+
+#[derive(Error, Debug, Clone, PartialEq, Eq)]
+#[error("Invalid external ID: {external_id} contains invalid character: {invalid_char}")]
+pub struct InvalidExternalIdError {
+    pub external_id: String,
+    pub invalid_char: char,
+}
+
+type ExternalIdMapperFn =
+    Box<dyn Fn(&EntityTopicId, &EntityExternalId) -> EntityExternalId + Send + Sync + 'static>;
+type ExternalIdValidatorFn =
+    Box<dyn Fn(&str) -> Result<EntityExternalId, InvalidExternalIdError> + Send + Sync + 'static>;
+
+/// Represents an error encountered while updating the store.
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("The specified entity {0} does not exist in the store")]
+    UnknownEntity(String),
+
+    #[error(transparent)]
+    InvalidExternalId(#[from] InvalidExternalIdError),
+
+    #[error("Auto registration of the entity with topic id {0} failed as it does not match the default topic scheme: 'device/<device-id>/service/<service-id>'. Try explicit registration instead.")]
+    NonDefaultTopicScheme(EntityTopicId),
+}
+
+pub(crate) struct EntityCache {
```

Comment:
A comment telling the purpose of this cache would be really helpful.

- What the difference of this cache compared to the entity store?
- How both cooperate?
- What's the content of the pending entities?

Reply:
Yes, I'll be adding a detailed comment for the `EntityCache` and the newly introduced `EntityMetadata` structs. Here is the quick summary though:

> What the difference of this cache compared to the entity store?

While `EntityStore`, owned by the agent, maintains **all** the metadata associated with an entity like its registration message, device twin data, supported operations etc, the `EntityCache` was envisaged as a lightweight map maintained by the mappers to maintain the external id mapping and a minimal set of additional metadata required to do the appropriate mapping to the cloud twins. 

> How both cooperate?

The entity cache is updated every time a new entity is successfully registered with the entity store(when an entity registration message is emitted either directly by the client or by the agent in response to a registrartion attempt over HTTP). When the mapper receives an entity registration message either way, it assumes that it was successfully registered and adds it to its entity cache.

> What's the content of the pending entities?

When the mapper receives an entity registration message with a parent, and if that parent is not registered yet at that point, it is added to the pending entity store and such pending entities are added to the entity cache only when their parent is registered. So, whenever a new entity is registered, the mapper checks its entity cache to find any of its children that arrived earlier, waiting to be registered and registers them along with the parent.